### PR TITLE
Multi level file dependency

### PR DESCRIPTION
**Pull request by ronykoz**
## Status
In Progress

## Related Issues
fixes: https://github.com/demisto/etc/issues/14683

## Description
This will make sure we run tests for all files that are dependant on the changes developers make

### CODE REVIEW

**Comment by ronykoz** [comment]
Before merging this issue we will have to re-run the creation of the id_set.json to make sure we don't exclude some id's behind.
I also thought about changing the template of the PR to include a place where we will copy-paste all the files that should be checked for the changes the developer is making, what do you think about that?

**Comment by ronykoz** [comment]
I probably should split the id_set into several files so the process will be more efficient, but the code can be reviewed, and it seems to be working

File: Tests/scripts/update_id_set.py [hunk 1]
```diff
@@ -109,89 +113,251 @@ def get_to_version(file_path):
...
+    cmd_list = []
+    data_dictionary = get_json(file_path)
+    commands = data_dictionary.get('script', {}).get('commands', [])
+    for command in commands:
+        cmd_list.append(command.get('name'))
+
+    return cmd_list
+
+
+def get_ids_from_playbook(param_to_enrich_by, data_dict):
```
Comment (ronykoz) [inline Tests/scripts/update_id_set.py]:
```suggestion
 def get_task_ids_from_playbook(param_to_enrich_by, data_dict):
```

File: Tests/scripts/update_id_set.py [hunk 2]
```diff
@@ -109,89 +113,251 @@ def get_to_version(file_path):
...
     print("Getting added files")
-    files_string = run_git_command("git diff --name-status {}".format(git_sha))
+    files_string = run_git_command("git diff --name-status HEAD")
     second_files_string = run_git_command("git diff --name-status origin/master...{}".format(branch_name))
-    added_files = get_added_files(files_string + '\n' + second_files_string)
+    added_files, modified_files = get_changed_files(files_string + '\n' + second_files_string)
 
-    if added_files:
+    if added_files or modified_files:
         print("Updating id_set.json")
```
Comment (ronykoz) [inline Tests/scripts/update_id_set.py]: this is not updating

File: Tests/scripts/update_id_set.py [hunk 3]
```diff
@@ -109,89 +113,251 @@ def get_to_version(file_path):
...
+            if re.match(INTEGRATION_REGEX, file_path, re.IGNORECASE):
+                integration_set.append(get_integration_data(file_path))
+                print("Adding {0} to id_set".format(get_script_or_integration_id(file_path)))
+            if re.match(SCRIPT_REGEX, file_path, re.IGNORECASE):
+                script_set.append(get_script_data(file_path))
+                print("Adding {0} to id_set".format(get_script_or_integration_id(file_path)))
+            if re.match(PLAYBOOK_REGEX, file_path, re.IGNORECASE):
+                playbook_set.append(get_playbook_data(file_path))
+                print("Adding {0} to id_set".format(collect_ids(file_path)))
+            if re.match(TEST_PLAYBOOK_REGEX, file_path, re.IGNORECASE):
```
Comment (ronykoz) [inline Tests/scripts/update_id_set.py]: what about test scripts?

File: Tests/scripts/update_id_set.py [hunk 4]
```diff
@@ -109,89 +113,251 @@ def get_to_version(file_path):
...
-                    "toversion": get_to_version(file_path)
-                }
-                id_dict = {
-                    id: versioning
-                }
-                id_list.append(id_dict)
-                print("Adding {0} to id_set".format(id))
-            if re.match(PLAYBOOK_REGEX, file_path, re.IGNORECASE) or re.match(TEST_PLAYBOOK_REGEX, file_path, re.IGNORECASE):
+                integration_data = get_integration_data(file_path)
+                integration_data_value = integration_data.values()[0]
```
Comment (ronykoz) [inline Tests/scripts/update_id_set.py]:
```suggestion
                 integration_data_value = integration_data[id]
```
same for all

File: Tests/scripts/update_id_set.py [hunk 5]
```diff
@@ -109,89 +113,251 @@ def get_to_version(file_path):
...
+                for integration in integration_set:
+                    integration_id = integration.keys()[0]
+                    if id == integration_id:
+                        integration[id] = integration_data_value
```
Comment (ronykoz) [inline Tests/scripts/update_id_set.py]: refer to versions (to, from)
here and everywhere

File: Tests/scripts/validate_files_structure.py [hunk 6]
```diff
@@ -351,47 +324,174 @@ def changed_context(file_path):
...
+
+    is_found = False
+    for playbook in playbook_set:
+        playbook_id = playbook.keys()[0]
+        if playbook_id == file_id:
+            is_found = True
+            if playbook != playbook_data:
+                print_error("You have failed to update id_set.json with the data of {} "
+                            "please run `python Tests/scripts/update_id_set.py`".format(file_path))
+                return False
```
Comment (ronykoz) [inline Tests/scripts/validate_files_structure.py]: need version support

File: Tests/scripts/validate_files_structure.py [hunk 7]
```diff
@@ -351,47 +324,174 @@ def changed_context(file_path):
...
+
+    if not is_found:
+        print_error("You have failed to update id_set.json with the data of {} "
+                    "please run `python Tests/scripts/update_id_set.py`".format(file_path))
+
+    return is_found
+
+
+def script_valid_in_id_set(file_path, script_set):
```
Comment (ronykoz) [inline Tests/scripts/validate_files_structure.py]: code duplication: pass the first line function as argument (or the result of that)

File: Tests/scripts/validate_files_structure.py [hunk 8]
```diff
@@ -351,47 +324,174 @@ def changed_context(file_path):
...
+    for file_path in added_files:
         print "Validating {}".format(file_path)
         if not validate_schema(file_path):
             has_schema_problem = True
-        if not is_release_branch() and not validate_file_release_notes(file_path):
-            has_schema_problem = True
 
-        if re.match(PLAYBOOK_REGEX, file_path, re.IGNORECASE) or re.match(TEST_PLAYBOOK_REGEX, file_path, re.IGNORECASE):
-            if changed_id(file_path):
+        if re.match(TEST_PLAYBOOK_REGEX, file_path, re.IGNORECASE):
```
Comment (ronykoz) [inline Tests/scripts/validate_files_structure.py]: check test script

File: Tests/scripts/validate_files_structure.py [hunk 9]
```diff
@@ -351,47 +324,174 @@ def changed_context(file_path):
...
+        if not is_release_branch() and not validate_file_release_notes(file_path):
             has_schema_problem = True
 
+        if re.match(PLAYBOOK_REGEX, file_path, re.IGNORECASE):
+            if is_circle and not playbook_valid_in_id_set(file_path, playbook_set):
+                has_schema_problem = True
+
         if re.match(TEST_PLAYBOOK_REGEX, file_path, re.IGNORECASE):
-            if not is_test_in_conf_json(file_path):
+            if is_circle and not playbook_valid_in_id_set(file_path, test_playbook_set):
```
Comment (ronykoz) [inline Tests/scripts/validate_files_structure.py]: test script

File: Tests/scripts/configure_tests.py [hunk 10]
```diff
@@ -161,70 +168,91 @@ def collect_tests(script_ids, playbook_ids, intergration_ids):
...
     catched_playbooks = set([])
     catched_intergrations = set([])
-    with open("./Tests/conf.json", 'r') as conf_file:
-        conf = json.load(conf_file)
 
-    conf_tests = conf['tests']
-    for t in conf_tests:
-        playbook_id = t['playbookID']
-        integrations_conf = t.get('integrations', [])
+    get_test_names(test_names)
```
Comment (ronykoz) [inline Tests/scripts/configure_tests.py]: don't do that!
return the list

File: Tests/scripts/configure_tests.py [hunk 11]
```diff
@@ -161,70 +168,91 @@ def collect_tests(script_ids, playbook_ids, intergration_ids):
...
+                catched_playbooks.add(playbook)
 
-            for task in tasks.values():
-                task_details = task.get('task', {})
+        if integration_to_command:
+            for command in test_playbook_data.get('implementing_commands', []):
+                for integration_id, integration_commands in integration_to_command.items():
+                    if command in integration_commands:
+                        tests.add(test_playbook_id)
+                        catched_playbooks.add(integration_id)
```
Comment (ronykoz) [inline Tests/scripts/configure_tests.py]: bug
```suggestion
                         catched_integrations.add(integration_id)
```

File: Tests/scripts/update_id_set.py [hunk 12]
```diff
@@ -109,89 +113,247 @@ def get_to_version(file_path):
...
+    return list(set([cmd.split('|')[-1] for cmd in depends_on]))
+
+
+def update_object_in_id_set(obj_id, obj_data, file_path, instances_set):
+    file_to_version = get_to_version(file_path)
+    file_from_version = get_from_version(file_path)
+
+    for instance in instances_set:
+        instance_id = instance.keys()[0]
+        integration_to_version = instance[instance_id].get('fromversion', '0.0.0')
```
Comment (yaakovi) [inline Tests/scripts/update_id_set.py]:
```suggestion
        integration_to_version = instance[instance_id].get('toversion', '99.99.99')
```

File: Tests/scripts/configure_tests.py [hunk 13]
```diff
@@ -247,13 +277,103 @@ def find_tests_for_modified_files(modified_files):
...
+    playbook_set = id_set['playbooks']
+    integration_set = id_set['integrations']
+
+    updated_script_names = set([])
+    updated_playbook_names = set([])
+
+    for script_id in script_names:
+        enrich_for_script_id(script_id, script_names, script_set, playbook_set, playbook_names, updated_script_names, updated_playbook_names)
+
+    integration_to_command = get_integration_commands(intergration_ids, integration_set)
```
Comment (anara123) [inline Tests/scripts/configure_tests.py]: intergration_ids => integration_ids

**Pull request merged**